homeassistant/helpers/storage.py

```diff
@@ -141,27 +152,38 @@ def async_delay_save(self, data_func: Callable[[], Dict], delay: float = 0) -> N
         self._data = {"version": self.version, "key": self.key, "data_func": data_func}
 
         self._async_cleanup_delay_listener()
+        self._async_cleanup_final_write_listener()
+
+        if self.hass.state == CoreState.stopping:
+            self._async_ensure_final_write_listener()
+            return
 
         self._unsub_delay_listener = async_call_later(
             self.hass, delay, self._async_callback_delayed_write
         )
 
-        self._async_ensure_stop_listener()
+    @callback
+    def _async_ensure_final_write_listener(self):
+        """Ensure that we write if we quit before delay has passed."""
+        if self._unsub_final_write_listener is None:
+            self._unsub_final_write_listener = self.hass.bus.async_listen_once(
+                EVENT_HOMEASSISTANT_FINAL_WRITE, self._async_callback_final_write
+            )
 
     @callback
     def _async_ensure_stop_listener(self):
         """Ensure that we write if we quit before delay has passed."""
         if self._unsub_stop_listener is None:
```

Comment:
Why would we still have a stop listener ? That should just become a final write listener ?

Reply:
removed... after staring at it for a while I got stuck in my head on something and dints see the easier solution!